src/controller.rs

```diff
@@ -45,7 +45,20 @@ impl<'b> Controller<'b> {
 
         let stdin = io::stdin();
 
-        for input_file in &self.config.files {
+        let filenames = if self.config.filenames.is_none() {
+            vec![None; self.config.files.len()]
+        } else {
+            self.config
+                .filenames
+                .as_ref()
+                .unwrap()
+                .into_iter()
+                .map(|name| Some(*name))
+                .collect()
+        };
```

Comment:
Similar here. Using a `match` statement would be more idiomatic and shorter (in general, we try to avoid `.unwrap` calls):
``` rust
let filenames = match self.config.filenames {
    Some(ref filenames) => filenames.into_iter().map(|name| Some(*name)).collect(),
    None => vec![None; self.config.files.len()],
};
```
Seeing that this vector is only being built up to be iterated over directly afterwards, we can also skip the `.collect()` part and just use the iterator directly:
``` rust
let filenames: Box<dyn Iterator<Item = _>> = match self.config.filenames {
    Some(ref filenames) => Box::new(filenames.into_iter().map(|name| Some(*name))),
    None => Box::new(std::iter::repeat(None)),
};
```

Reply:
Cool! I need to read up on the Box more. Why use it here? Is it for performance?

Reply:
`Box` is used for something different here.

We want `filenames` to be an iterator over `&str` (`Iterator<Item = &str>`). However, `Iterator` is a `trait` and not a *concrete type*. It's more like an interface (Java) or an abstract base class (C++, Python, ...). There are many different concrete types that *implement* `Iterator<Item = &str>`. In our case, the concrete types are `std::iter::Map<I, F>` (which is the return type of `.map(…)`) and `std::iter::Repeat<A>` (the return type of `iter::repeat`).

These two types are completely different (i.e. have a different memory layout and size). Consequently, there is no way to store simply say
``` rust
// does not compile:
let filenames = match self.config.filenames {
    Some(ref filenames) => filenames.into_iter().map(|name| Some(*name)),
    None => std::iter::repeat(None),
};
```
because we cannot give a *statically known* type to `filenames`.

The way to solve this problem is to use dynamic memory allocation. Instead of creating the `filenames` object on the stack (where we would need to know the static size), we dynamically allocate space for it on the heap. This is what `Box` does. It's a smart pointer that references a dynamically sized object on the heap.

Using `Box` also allows us to abstract over the concrete type in use. This is where the type annotation comes into play:
``` rust
let filenames: Box<dyn Iterator<Item = _>> = …
```
This basically says: `filenames` is a smart pointer to an object of some type that *implements* `Iterator<…>`. We don't have to specify what `Item` is here, because that can be inferred by the compiler.

If you are familiar with C++, `Box` is like a (smart) pointer to the abstract base class.

Reply:
Whoa thanks for the explanation! I am familiar with C++ and smart pointers so this makes a lot of sense now. Thank you !